pyasn1/codec/ber/decoder.py

```diff
@@ -1447,313 +1631,336 @@ def __call__(self, substrate, asn1Spec=None,
 
-                    if isShortTag:
-                        # cache short tags
-                        tagCache[firstOctet] = lastTag
+        while True:
 
-                if tagSet is None:
-                    if isShortTag:
-                        try:
-                            tagSet = tagSetCache[firstOctet]
+            if LOG:
+                LOG('decoder called at scope %s with state %d, working with up '
+                    'to %s octets of substrate: '
+                    '%s' % (debug.scope, state, length, substrate))
 
-                        except KeyError:
-                            # base tag not recovered
-                            tagSet = tag.TagSet((), lastTag)
-                            tagSetCache[firstOctet] = tagSet
-                    else:
-                        tagSet = tag.TagSet((), lastTag)
+            # Look for end-of-octets sentinel
+            if allowEoo and self.supportIndefLength:
+
+                for eoo_candidate in _read(substrate, 2):
+                    if eoo_candidate is None:
+                        yield
+
+                if eoo_candidate == self.__eooSentinel:
+                    if LOG:
+                        LOG('end-of-octets sentinel found')
+                    yield eoo.endOfOctets
+                    return
 
                 else:
-                    tagSet = lastTag + tagSet
+                    substrate.seek(-2, os.SEEK_CUR)
 
-                state = stDecodeLength
+            value = noValue
 
-                if LOG:
-                    LOG('tag decoded into %s, decoding length' % tagSet)
+            substrate._markedPosition = substrate.tell()
 
-            if state is stDecodeLength:
-                # Decode length
-                try:
-                    firstOctet = ord(substrate.read(1))
-                except:
-                    raise error.SubstrateUnderrunError(
-                        'Short octet stream on length decoding'
-                    )
+            while state is not stStop:
 
-                if firstOctet < 128:
-                    length = firstOctet
+                if state is stDecodeTag:
+                    # Decode tag
+                    isShortTag = True
 
-                elif firstOctet > 128:
-                    size = firstOctet & 0x7F
-                    # encoded in size bytes
-                    encodedLength = list(substrate.read(size))
-                    # missing check on maximum size, which shouldn't be a
-                    # problem, we can handle more than is possible
-                    if len(encodedLength) != size:
-                        raise error.SubstrateUnderrunError(
-                            '%s<%s at %s' % (size, len(encodedLength), tagSet)
+                    for firstByte in _read(substrate, 1):
+                        if firstByte is None:
+                            yield
+
+                    if not firstByte:
+                        yield
```

Comment:
I'd guess this (we read zero bytes) is semantically different from `firstByte` being None (bytes not available). Should we `return` instead?

Reply:
Actually, this cannot happen as a zero-length bytes is never yielded from _read.